File: AI/app/blueprints/llm_deploy.py
"""
GPU 节点 Qwen 大模型部署 API。
"""
import logging

from flask import Blueprint, jsonify, request

from app.services.llm_deploy_service import (
    check_llm_deploy_vram,
    delete_llm_deploy,
    deploy_llm,
    get_llm_catalog,
    list_deployed_llm_services,
    receive_llm_heartbeat,
    stop_llm_deploy,
)

llm_deploy_bp = Blueprint('llm_deploy', __name__)
logger = logging.getLogger(__name__)


@llm_deploy_bp.route('/catalog', methods=['GET'])
def catalog_route():
    return jsonify(get_llm_catalog())


@llm_deploy_bp.route('/list', methods=['GET'])
def list_route():
    page = request.args.get('page', 1, type=int)
    page_size = request.args.get('pageSize', 20, type=int)
    return jsonify(list_deployed_llm_services(page, page_size))


@llm_deploy_bp.route('/check-vram', methods=['GET'])
def check_vram_route():
    try:
        qwen_model_key = request.args.get('qwen_model_key', '')
        target_node_id = request.args.get('target_node_id', type=int)
        tensor_parallel_size = request.args.get('tensor_parallel_size', 1, type=int)
        if not qwen_model_key:
            return jsonify({'code': 400, 'msg': '缺少 qwen_model_key'}), 400
        if not target_node_id:
            return jsonify({'code': 400, 'msg': '缺少 target_node_id'}), 400
        return jsonify(
            check_llm_deploy_vram(
                qwen_model_key,
                target_node_id=target_node_id,
                tensor_parallel_size=tensor_parallel_size,
            )
        )
    except ValueError as exc:
        return jsonify({'code': 400, 'msg': str(exc)}), 400
    except Exception as exc:
        logger.exception('显存预检失败')
        return jsonify({'code': 500, 'msg': str(exc)}), 500


@llm_deploy_bp.route('/deploy', methods=['POST'])
def deploy_route():
    try:
        data = request.get_json() or {}
        qwen_model_key = data.get('qwen_model_key')
        if not qwen_model_key:
            return jsonify({'code': 400, 'msg': '缺少 qwen_model_key'}), 400
        result = deploy_llm(
            qwen_model_key,
            start_port=int(data.get('start_port', 8100)),
            target_node_id=int(data['target_node_id']) if data.get('target_node_id') else None,
            auto_schedule=bool(data.get('auto_schedule', False)),
            tensor_parallel_size=int(data.get('tensor_parallel_size', 1)),
            max_model_len=int(data.get('max_model_len', 8192)),
            service_name=data.get('service_name'),
        )
        return jsonify(result)
    except ValueError as exc:
        return jsonify({'code': 400, 'msg': str(exc)}), 400
    except Exception as exc:
        logger.exception('LLM 部署失败')
        return jsonify({'code': 500, 'msg': str(exc)}), 500


@llm_deploy_bp.route('/heartbeat', methods=['POST'])
def heartbeat_route():
    try:
        data = request.get_json() or {}
        return jsonify(receive_llm_heartbeat(data))
    except ValueError as exc:
        return jsonify({'code': 400, 'msg': str(exc)}), 400
    except Exception as exc:
        logger.exception('LLM 心跳处理失败')
        return jsonify({'code': 500, 'msg': str(exc)}), 500


@llm_deploy_bp.route('/stop', methods=['POST'])
def stop_route():
    try:
        data = request.get_json() or {}
        service_id = data.get('service_id')
        if not service_id:
            return jsonify({'code': 400, 'msg': '缺少 service_id'}), 400
        return jsonify(stop_llm_deploy(int(service_id)))
    except ValueError as exc:
        return jsonify({'code': 400, 'msg': str(exc)}), 400
    except Exception as exc:
        logger.exception('停止 LLM 部署失败')
        return jsonify({'code': 500, 'msg': str(exc)}), 500


@llm_deploy_bp.route('/delete', methods=['DELETE'])
def delete_route():
    try:
        service_id = request.args.get('id', type=int)
        if not service_id:
            return jsonify({'code': 400, 'msg': '缺少 id'}), 400
        return jsonify(delete_llm_deploy(service_id))
    except ValueError as exc:
        return jsonify({'code': 400, 'msg': str(exc)}), 400
    except Exception as exc:
        logger.exception('删除 LLM 部署失败')
        return jsonify({'code': 500, 'msg': str(exc)}), 500

File: WEB/src/api/device/llmDeploy.ts
import { defHttp } from '@/utils/http/axios';

enum Api {
  CATALOG = '/model/llm_deploy/catalog',
  LIST = '/model/llm_deploy/list',
  CHECK_VRAM = '/model/llm_deploy/check-vram',
  DEPLOY = '/model/llm_deploy/deploy',
  STOP = '/model/llm_deploy/stop',
  DELETE = '/model/llm_deploy/delete',
}

export interface QwenModelPreset {
  key: string;
  label: string;
  hf_model_id: string;
  model_type: string;
  min_vram_gb: number;
  recommended_gpu_count: number;
  max_model_len_default: number;
  description?: string;
}

export interface LLMDeployService {
  id?: number;
  service_name: string;
  qwen_model_key: string;
  hf_model_id: string;
  node_id?: number;
  server_ip?: string;
  port?: number;
  api_endpoint?: string;
  tensor_parallel_size?: number;
  max_model_len?: number;
  status?: string;
  llm_config_id?: number;
  process_id?: number;
  log_path?: string;
  error_message?: string;
  deploy_time?: string;
  last_heartbeat?: string;
  created_at?: string;
  updated_at?: string;
}

export interface LLMDeployParams {
  qwen_model_key: string;
  target_node_id?: number;
  auto_schedule?: boolean;
  start_port?: number;
  tensor_parallel_size?: number;
  max_model_len?: number;
  service_name?: string;
}

export function getLlmDeployCatalog() {
  return defHttp.get<{ data: QwenModelPreset[] }>({ url: Api.CATALOG });
}

export function getLlmDeployList(params?: { page?: number; pageSize?: number }) {
  return defHttp.get<{ data: { list: LLMDeployService[]; total: number } }>({
    url: Api.LIST,
    params,
  });
}

export interface LLMVramCheckResult {
  ok: boolean;
  model_key?: string;
  model_label?: string;
  node_id?: number;
  node_name?: string;
  node_host?: string;
  tensor_parallel_size?: number;
  required_vram_gb?: number;
  required_gpu_count?: number;
  free_vram_gb?: number;
  selected_free_vram_gb?: number;
  total_vram_gb?: number;
  gpu_count?: number;
  gpus?: Array<{
    id?: number;
    name?: string;
    mem_total_gb?: number;
    mem_free_gb?: number;
    util?: number;
  }>;
  selected_gpus?: Array<{
    id?: number;
    name?: string;
    mem_free_gb?: number;
    mem_total_gb?: number;
  }>;
}

export function checkLlmDeployVram(params: {
  qwen_model_key: string;
  target_node_id: number;
  tensor_parallel_size?: number;
}) {
  return defHttp.get<{ msg: string; data: LLMVramCheckResult }>({
    url: Api.CHECK_VRAM,
    params,
  });
}

export function deployLlmModel(data: LLMDeployParams) {
  return defHttp.post<{ data: LLMDeployService; msg: string }>({ url: Api.DEPLOY, data });
}

export function stopLlmDeploy(serviceId: number) {
  return defHttp.post({ url: Api.STOP, data: { service_id: serviceId } });
}

export function deleteLlmDeploy(id: number) {
  return defHttp.delete({ url: Api.DELETE, params: { id } });
}
